src/core/ChannelHandler.js

```diff
@@ -0,0 +1,109 @@
+import { Observable } from 'rxjs';
+import debug from 'debug';
+import assert from 'assert';
+
+const createLogger = label => debug(`log-${label}`);
+
+const events = [
+  'phx_reply',
+  'phx_error',
+  'status_changed',
+  'error',
+  'failed',
+  'starting',
+  'started',
+  'deploying',
+  'deployed',
+  'deployment_failed',
+  'ready',
+  'terminated',
+  'snapshot_taken',
+  'snapshot_reverted'
+];
+
+export default class ChannelHandler {
+  constructor(name, socket) {
+    this._name = name;
+    this._channel = socket.channel(this._name);
+    this._stream = this._buildChannelStream(events);
+    this._log = createLogger(this._name);
+    this._logger = this._stream.subscribe(value =>
+      this._log(JSON.stringify(value, null, 4))
+    );
+    this.init();
+  }
+
+  _buildChannelStream(eventsList) {
+    return new Observable(subscriber => {
+      const setChannelEvent = event => {
+        this._channel.on(event, data => {
+          switch (event) {
+            case 'deployment_failed':
+              subscriber.error({ event, payload: data });
+              break;
+            case 'error':
+              subscriber.error({ event, payload: data });
+              break;
+            case 'failed':
+              subscriber.error({ event, payload: data });
+              break;
+            default:
+              subscriber.next({ event, payload: data });
+          }
+        });
+      };
+
+      eventsList.forEach(event => {
+        setChannelEvent(event);
+      });
+    });
+  }
+
+  once(predicate) {
+    // name 'once' may be ambiguous
```

Comment:
can you elaborate?

Reply:
I think it is tied to the original `once()` function, and maybe where some of our confusion was originally. I want this function to resolve when an event has been fired that passes the predicate condition. `until()` sounds better for me because we would use it like 

```javascript
const res = await until(id, Event.CHAIN_STARTED);
```
But I could be getting bogged down in the linguistics